### DISCUSSION TITLE
How to build image with modified kernel source code?

### DISCUSSION THREAD

**Discussion by Ruach** [Joshua-Riek/ubuntu-rockchip#844] [Q&A] [answered]
Hi all, 

I wanted to modify the build/linux-rockchip source code to modify linux kernel.
After the modification, how to generate the image from the modified linux kernel image?

It seems that the kernel image is located in build/ubuntu-live-build/build/chroot/boot,
so it seems that the updated kernel image should be placed in the above path.

As I am pretty new to debian packaging, I am having a hard time to understand how the rootfile system is generated with ubuntu-live-build. Which file should I modify to make the generated rootfs will have the updated kernel image..?

**Comment by Joshua-Riek** [comment, marked as answer]
Sadly, part of following the upstream Ubuntu process makes changes like this difficult, there is pretty much no updated or valid documentation. Most of what I have done has been reverse engineered with hacks on top. The ubuntu-live-build repo will be deleted and merged into ubuntu-rockchip. I'm very burnt out from the Ubuntu 24.04 release and don't have time to write up a guide, the entire build process is not friendly to users or companies who want their own modifications because of the use of Launchpad.

But you can try this, I added a quick hack to let people define a kernel repo and branch during the build process if `--launchpad` is not passed during the build process.
https://github.com/Joshua-Riek/ubuntu-rockchip/blob/d4440960784cdfa63b8296aebf356cc46e383b5d/config/suites/noble.sh#L6-L7

**Reply by Ruach** [reply]
@Joshua-Riek 
Thanks for the reply. I still don't understand how the export KERNEL_REPO can be used to build ubuntu with custom built kernel. 

It seems that ubuntu-live-build automatically include linux 6.1.0.1014.14 for rockchip when it builds rootfs and image. 

## This is part of the configuration file locally generated
```
#
# Automatically generated file; DO NOT EDIT.
# Linux/arm64 6.1.43 Kernel Configuration
#
CONFIG_CC_VERSION_TEXT="aarch64-linux-gnu-gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0"
CONFIG_CC_IS_GCC=y
CONFIG_GCC_VERSION=110400
CONFIG_CLANG_VERSION=0
CONFIG_AS_IS_GNU=y
CONFIG_AS_VERSION=23800
CONFIG_LD_IS_BFD=y
CONFIG_LD_VERSION=23800
```

## And this is the configuration file located in chroot/boot
```
#
# Automatically generated file; DO NOT EDIT.
# Linux/arm64 6.1.0-1014-rockchip Kernel Configuration
#
CONFIG_CC_VERSION_TEXT="aarch64-linux-gnu-gcc (Ubuntu 13.2.0-23ubuntu4) 13.2.0"
CONFIG_CC_IS_GCC=y
CONFIG_GCC_VERSION=130200
CONFIG_CLANG_VERSION=0
CONFIG_AS_IS_GNU=y
CONFIG_AS_VERSION=24200
CONFIG_LD_IS_BFD=y
CONFIG_LD_VERSION=24200
```

Then what is the purpose of compiling linux kernel locally? For example, local linux compilation generates below two output files, but seems that it has not been used.. Is there any way to change config file to make live build use the deb files generated as a result of custom linux compilation instead of remote pre-compiled linux? 

+ linux-headers-6.1.43-rockchip_6.1.43-rockchip-1_arm64.deb  
+ linux-image-6.1.43-rockchip_6.1.43-rockchip-1_arm64.deb

**Reply by Joshua-Riek** [reply]
If I'm being honest, I want to remove building the kernel entirely during the build process and not even give an option for it. As the kernel should be packaged on Launchpad, however this is complicated, confuse anyone who wants their own images built, and is too much for me to document right now. 

So, there is a hack in config-image.sh to uninstall the kernel and re-install with the local kernel. I use this process when testing kernel changes and provisioning support for new boards.

**Reply by Ruach** [reply]
@Joshua-Riek Thanks a lot for the clarification. 

I have one more question: Is it possible to update the kernel with custom build linux-image.deb after booting up the orangepi with the ubuntu? 

After booting the orangepi-5-plus with ubuntu automatically built with launchpad, I installed the custom built deb package.
> $ dpkg -i linux-image-6.1.43-rockchip_6.1.43-rockchip-1_arm64.deb 

The dpkg was successfully installed, but the orangepi doesn't boot up after that. Is it a wrong way to install the custom built linux kernel on orangepi with ubuntu..? I booted the orangepi from sd card

**Comment by ohyeah521** [comment]
I also encountered the same problem, the 2.x version of the build system was not very friendly to custom kernels and did not find a solution, so I rolled back to version 1.33 and continued to use Ubuntu 22.04

**Reply by Ruach** [reply]
@ohyeah521  could you elaborate little more? When I roll back to version 1.33 does it allow me to compile ubuntu with custom built linux image? And does it also support orangepi-5-plus by any chance..?
